buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/CopyRestApiTask.java

```diff
@@ -0,0 +1,218 @@
+ * declared through the 'copyRestApiSpecs' extension. <br>
+ * This task supports copying only a subset of the Rest API specification through the use of the custom extension.<br>
+ * <i>For example:</i>
+ * <pre>
+ * copyRestApiSpecs {
+ *   includeXpack 'enrich'
+ * }
+ * </pre>
+ * Will copy any of the the x-pack specs that start with enrich*. The core API specs will also be copied iff the project also has
+ * Rest YAML tests. To help optimize the build cache, it is recommended to explicitly declare which specs your project depends on.
+ * <i>For example:</i>
+ * <pre>
+ * copyRestApiSpecs {
+ *   includeCore 'index', 'cat'
+ *   includeXpack 'enrich'
+ * }
+ *  </pre>
+ * <br>
+ * <strong>Rest YAML tests :</strong> <br>
+ * When the {@link CopyRestApiPlugin} has been applied this task can copy the Rest YAML tests iff explicitly configured with
+ * `includeCore` or `includeXpack` through the `copyYamlTests` extension.
+ * <i>For example:</i>
+ * <pre>
+ * copyYamlTests {
+ *   includeXpack 'graph'
+ * }
+ * </pre>
+ * Will copy any of the the x-pack tests that start with graph.
+ */
+public class CopyRestApiTask extends DefaultTask {
+    private static final Logger logger = Logging.getLogger(CopyRestApiTask.class);
+    final ListProperty<String> includeCore = getProject().getObjects().listProperty(String.class);
+    final ListProperty<String> includeXpack = getProject().getObjects().listProperty(String.class);
+
+    Configuration coreConfig;
+    Configuration xpackConfig;
+    String copyTo;
+
+    private final PatternFilterable corePatternSet;
+    private final PatternFilterable xpackPatternSet;
+
+    public CopyRestApiTask() {
+        corePatternSet = getPatternSetFactory().create();
+        xpackPatternSet = getPatternSetFactory().create();
+    }
+
+    @Inject
+    protected Factory<PatternSet> getPatternSetFactory() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Input
+    public ListProperty<String> getIncludeCore() {
+        return includeCore;
+    }
+
+    @Input
+    public ListProperty<String> getIncludeXpack() {
+        return includeXpack;
+    }
+
+    @SkipWhenEmpty
+    @InputFiles
+    public FileTree getInputDir() {
+        xpackPatternSet.setIncludes(includeXpack.get().stream().map(prefix -> prefix + "*/**").collect(Collectors.toList()));
+        ConfigurableFileCollection fileCollection = getProject().files(xpackConfig.getAsFileTree().matching(xpackPatternSet));
+        if (BuildParams.isInternal()) {
+            corePatternSet.setIncludes(includeCore.get().stream().map(prefix -> prefix + "*/**").collect(Collectors.toList()));
+            fileCollection.plus(coreConfig.getAsFileTree().matching(corePatternSet));
+        } else {
+            fileCollection.plus(coreConfig);
+        }
+        return "spec".equals(getTestOrSpec()) && projectHasYamlRestTests()
```

Comment:
I don't know how many times I've looked at this I still have trouble following this. Why the different behavior based on whether we are copying specs or tests? Also, as mentioned below, conditioning on whether the project "has yaml tests" should live in the plugin, not the task implementation. If this stuff is n/a for projects w/o yaml tests, then we can ditch creating this task altogether in the plugin itself.

Reply:
The only difference in behavior is if this is copying rest specs it should take factor in if the project has rest tests, else for tests it relies only the existence of the includes. Will move the has yaml tests check to the plugin and avoid the check here.

Reply:
So in the case that this is the copy specs task, and my project defines yaml tests, but no explicit incudes, what gets copied?

Reply:
the core specs will get copied.

Reply:
I found another case where I need to change behavior based on spec or tests.  Rather then some conditional logic from 1 task, I will split this task up into 2 tasks.